pkg/controllers/node/node_controller.go

```diff
@@ -193,3 +230,207 @@ func kick(c chan<- interface{}) {
+	expectedIPs := []string{}
+	if node.Spec.BGP != nil {
+		// BGP IPv4 and IPv6 addresses are CIDRs.
+		if node.Spec.BGP.IPv4Address != "" {
+			ip, _, _ := net.ParseCIDROrIP(node.Spec.BGP.IPv4Address)
+			expectedIPs = append(expectedIPs, ip.String())
+		}
+		if node.Spec.BGP.IPv6Address != "" {
+			ip, _, _ := net.ParseCIDROrIP(node.Spec.BGP.IPv6Address)
+			expectedIPs = append(expectedIPs, ip.String())
+		}
+		if node.Spec.BGP.IPv4IPIPTunnelAddr != "" {
+			expectedIPs = append(expectedIPs, node.Spec.BGP.IPv4IPIPTunnelAddr)
+		}
+	}
+	if node.Spec.IPv4VXLANTunnelAddr != "" {
+		expectedIPs = append(expectedIPs, node.Spec.IPv4VXLANTunnelAddr)
+	}
+	return expectedIPs
+}
+
+// generateHostendpointFromNode returns the expected auto hostendpoint to be
+// created from the given node.
+func (c *NodeController) generateHostendpointFromNode(node *api.Node) *api.HostEndpoint {
+	hepLabels := make(map[string]string, len(node.Labels)+1)
+	for k, v := range node.Labels {
+		hepLabels[k] = v
+	}
+	hepLabels[hepCreatedLabelKey] = hepCreatedLabelValue
+
+	return &api.HostEndpoint{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   c.generateHostendpointName(node.Name),
+			Labels: hepLabels,
+		},
+		Spec: api.HostEndpointSpec{
+			Node:          node.Name,
+			InterfaceName: "*",
+			ExpectedIPs:   c.getHostendpointExpectedIPs(node),
+		},
+	}
+}
+
+// hostendpointNeedsUpdate returns true if the current automatic hostendpoint
+// needs to be updated.
+func (c *NodeController) hostendpointNeedsUpdate(current *api.HostEndpoint, expected *api.HostEndpoint) bool {
+	if !reflect.DeepEqual(current.Labels, expected.Labels) {
+		return true
+	}
+	if !reflect.DeepEqual(current.Spec.ExpectedIPs, expected.Spec.ExpectedIPs) {
+		return true
+	}
+	return current.Spec.InterfaceName != expected.Spec.InterfaceName
+}
+
+// updateHostendpoint updates the current hostendpoint so that it matches the
+// expected hostendpoint.
+func (c *NodeController) updateHostendpoint(current *api.HostEndpoint, expected *api.HostEndpoint) error {
+	if c.hostendpointNeedsUpdate(current, expected) {
+		log.WithField("hostendpoint", current.Name).Debug("hostendpoint needs update")
+		expected.ResourceVersion = current.ResourceVersion
+		expected.ObjectMeta.CreationTimestamp = current.ObjectMeta.CreationTimestamp
+		expected.ObjectMeta.UID = current.ObjectMeta.UID
+
+		time.Sleep(c.rl.When(RateLimitCalicoUpdate))
+		_, err := c.calicoClient.HostEndpoints().Update(context.Background(), expected, options.SetOptions{})
+		if err == nil {
+			c.rl.Forget(RateLimitCalicoUpdate)
+		}
+		return err
+	}
+	return nil
+}
+
+// syncHostendpoint syncs the auto hostendpoint for the given node.
+func (c *NodeController) syncHostendpoint(node *api.Node) error {
+	hepName := c.generateHostendpointName(node.Name)
+	log.Debugf("syncing hostendpoint %q from node %+v", hepName, node)
+
+	// On failure, we retry a certain number of times.
+	for n := 0; n < 5; n++ {
```

Comment:
I think we should move the retry logic to the top level, rather than tightly wrap each datastore operation.

If an individual data store operation fails, we want to start the entire thing from the beginning since our state might be out of date. For example, we might List() all HEPs, and then fail to Update() one, in which case we want to List() them all again to make sure we've got the latest state (in case one was modified in between the List() and Update() calls).

Reply:
Good point, I addressed this. LMK what you think about the approach. Functions like `deleteHostendpoint` are used at the "outer" level and thus would use retries but also within functions that already have retries (like `syncAllAutoHostendpoints`)